### PR TITLE
feat(rspack): add generatePackageJson plugin

### LINKED ISSUE #226: Add webpack's generatePackageJson option to rspack's project configuration
### What problem does this feature solve?

The generatePackageJson option can be configured in project.json file and consumed by the Webpack.  It's desirable that this option is supported by Rspack.

**generatePackageJson**
Generates a package.json and pruned lock file with the project's node_module dependencies populated for installing in a container. If a package.json exists in the project's directory, it will be reused with dependencies populated.

### What does the proposed API of configuration look like?

```json
         "options": {
              ...,
              "maxWorkers": 1,
              "memoryLimit": 2048,
              "generatePackageJson": false,
              ...,
         }
```

### PR DESCRIPTION
**Pull request by travtarr**
Closes #226

### CODE REVIEW

File: packages/rspack/src/plugins/generate-package-json-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import {
+  HelperDependency,
+  createLockFile,
+  createPackageJson,
+  getHelperDependenciesFromProjectGraph,
+  getLockFileName,
+  readTsConfig,
+} from '@nx/js';
+import { type Compiler, type RspackPluginInstance } from '@rspack/core';
+import { RawSource } from 'webpack-sources';
```
Comment (travtarr) [inline packages/rspack/src/plugins/generate-package-json-plugin.ts:16]: This is what the underlying `@rspack/core` package uses as there is no `rspack-sources` yet.

File: packages/rspack/src/plugins/generate-package-json-plugin.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+const pluginName = 'GeneratePackageJsonPlugin';
+
+export class GeneratePackageJsonPlugin implements RspackPluginInstance {
```
Comment (travtarr) [inline packages/rspack/src/plugins/generate-package-json-plugin.ts:20]: This is essentially identical to `@nx/webpack`'s `GeneratePackageJsonPlugin` but utilizing the `@rspack/core` dependencies.

https://github.com/nrwl/nx/blob/e29e9f90b6c5cf2a5b686327d68de15b873c7c4c/packages/webpack/src/plugins/generate-package-json-plugin.ts

File: packages/rspack/src/utils/with-nx.ts [hunk 3]
```diff
@@ -37,10 +43,25 @@ export function withNx(_opts = {}) {
       );
     }
 
-    options.fileReplacements.forEach(item => {
+    if (options.generatePackageJson) {
+      const mainOutputFile =
+        options.main.split('/').pop().split('.')[0] + '.js';
```
Comment (travtarr) [inline packages/rspack/src/utils/with-nx.ts:48]: This is a bit ugly so I'm open to alternatives.

**Comment by travtarr** [comment]
The e2e tests pass and I've successfully ran the builds when linking this to all my other Nest projects utilizing this rspack executor.

**Comment by kodeine** [comment]
@mandarini whenever you get time, can we please merge? thank you

**Review by mandarini — APPROVED** [review]
I'm approving @travtarr @kodeine , but please can you rebase with latest `main`?

**Comment by travtarr** [comment]
> I'm approving @travtarr @kodeine , but please can you rebase with latest `main`?

It's been rebased, thanks!

**Comment by kodeine** [comment]
@mandarini sorry for the tag :) would really appreciate if we can merge this, thank you for all the work you do!

**Pull request merged**